jedeschule/spiders/berlin.py

```diff
@@ -1,102 +1,84 @@
-                             meta=meta, callback=self.call_students, dont_filter=True)
-
-    def call_students(self, response):
-        links = response.css('#NaviSchuelerschaft ul ul li a::attr(href)').extract()
-        meta = response.meta
-        meta['cookiejar'] = response.meta['cookiejar']
-        if len(links) > 0:
-            meta['student_links'] = links[1:]
-
-            yield scrapy.Request(self.base_url + links[0], meta=response.meta, dont_filter=True, callback=self.parse_students)
-        else:
-            yield response.meta
+        yield meta
 
-    def parse_students(self, response):
-        # inspect_response(response, self)
-        year = response.request.url.rsplit('=', 1)[1]
-        meta = response.meta
-        if 'students' not in meta.keys():
-            meta['students'] = {}
-        meta['students'][year] = self.parse_table(response.css('#ContentPlaceHolderMenuListe_GridViewJahrgansstufen'))
-        links = meta['student_links']
-
-        if len(links) > 0:
-            meta['student_links'] = links[1:]
-            yield scrapy.Request(self.base_url + links[0], meta=response.meta, dont_filter=True, callback=self.parse_students)
-        else:
-            yield scrapy.Request('https://www.berlin.de/sen/bildung/schule/berliner-schulen/schulverzeichnis/schulpersonal.aspx?view=pers',
-                                 meta=meta, callback=self.call_teacher, dont_filter=True)
-
-    def call_teacher(self, response):
-        links = response.css('#NaviSchulpersonal ul ul li a::attr(href)').extract()
-        meta = response.meta
-        meta['cookiejar'] = response.meta['cookiejar']
-        if len(links) > 0:
-            meta['teacher_links'] = links[1:]
-            yield scrapy.Request(self.base_url + links[0], meta=response.meta, dont_filter=True, callback=self.parse_teachers)
-        else:
-            yield response.meta
+    def _parse_school_no(self, url):
+        """Parses the school number from the 'IDSchulzweig' parameter in the url"""
+        parsed = urlparse.urlparse(url)
+        id_in_url: List[str] = parse_qs(parsed.query, max_num_fields=1)['IDSchulzweig']
+        assert len(id_in_url) == 1
 
-    def parse_teachers(self, response):
-        year = response.request.url.rsplit('=', 1)[1]
-        meta = response.meta
-        if 'teachers' not in meta.keys():
-            meta['teachers'] = {}
-        meta['teachers'][year] = self.parse_table(response.css('#ContentPlaceHolderMenuListe_GridViewPersonal'))
-        links = meta['teacher_links']
+        return id_in_url[0].strip()
 
-        if len(links) > 0:
-            meta['teacher_links'] = links[1:]
-            yield scrapy.Request(self.base_url + links[0], meta=meta, dont_filter=True, callback=self.parse_students)
-        else:
-            # inspect_response(response, self)
-            yield meta
+    # fix wrong tabs, spaces and new lines
+    def fix_data(self, string):
```

Comment:
Pudo once wrote a little python library called [normality](https://github.com/pudo/normality) to handle string cleaning like this. We could use the `collapse_spaces` function from there instead. This is not a big gain though unless this cleaning is also going to be required in other scrapers.

Reply:
Let's keep that as it is for now, but keep normality in mind for later.